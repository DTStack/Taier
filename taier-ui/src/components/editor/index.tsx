/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import type { CSSProperties } from 'react';
import { useEffect, useRef } from 'react';
import * as monaco from '@dtinsight/molecule/esm/monaco';
import classNames from 'classnames';
import { defaultOptions } from './config';
import './language/jsonlog';
import './style.scss';

interface IEditorProps {
    className?: string;
    style?: CSSProperties;
    value?: string;
    language?: string;
    sync?: boolean;
    options?: monaco.editor.IStandaloneEditorConstructionOptions;
    cursorPosition?: monaco.IPosition;
    placeholder?: string;
    onChange?: (value: string, instance: monaco.editor.IStandaloneCodeEditor) => void;
    onBlur?: (value: string) => void;
    onFocus?: (value: string) => void;
    onCursorSelection?: (value: string) => void;
}

export default function Editor({
    className,
    style,
    options,
    value,
    sync,
    language,
    cursorPosition,
    placeholder,
    onChange,
    onBlur,
    onFocus,
    onCursorSelection,
}: IEditorProps) {
    const container = useRef<HTMLDivElement>(null);
    const monacoEditor = useRef<monaco.editor.IStandaloneCodeEditor>();
    const placeholderDOM = useRef<HTMLPreElement>(null);

    const initMonaco = () => {
        if (container.current) {
            const editorOptions = {
                ...defaultOptions,
                ...options,
                value,
                language: language || 'sql',
            };

            monacoEditor.current = monaco.editor.create(container.current, editorOptions);

            handleShowPlaceholder(value);

            if (monacoEditor.current && cursorPosition) {
                monacoEditor.current.setPosition(cursorPosition);
                monacoEditor.current.focus();
                monacoEditor.current.revealPosition(cursorPosition, monaco.editor.ScrollType.Immediate);
            }

            initEditor();
            registerCommands();
        }
    };

    /**
     * TODO: I don't why lost these commands, it should be figure it out after a while
     */
    const registerCommands = () => {
        if (monacoEditor.current) {
            monacoEditor.current.addCommand(monaco.KeyMod.CtrlCmd | monaco.KeyCode.KeyZ, () => {
                monacoEditor.current?.trigger('editor', 'undo', null);
            });

            monacoEditor.current.addCommand(monaco.KeyMod.CtrlCmd | monaco.KeyCode.KeyA, () => {
                monacoEditor.current?.trigger('editor', 'editor.action.selectAll', null);
            });

            monacoEditor.current.addCommand(monaco.KeyMod.CtrlCmd | monaco.KeyMod.Shift | monaco.KeyCode.KeyA, () => {
                monacoEditor.current?.trigger('editor', 'redo', null);
            });

            // disabled these global commands in editor
            monacoEditor.current.addCommand(monaco.KeyMod.CtrlCmd | monaco.KeyCode.KeyF, () => {});
            monacoEditor.current.addCommand(monaco.KeyMod.CtrlCmd | monaco.KeyCode.KeyK, () => {});
            monacoEditor.current.addCommand(monaco.KeyMod.CtrlCmd | monaco.KeyCode.Comma, () => {});
            monacoEditor.current.addCommand(
                monaco.KeyMod.CtrlCmd | monaco.KeyMod.Shift | monaco.KeyCode.KeyP,
                () => {}
            );
        }
    };

    const initEditorEvent = () => {
        if (monacoEditor.current) {
            monacoEditor.current.onDidChangeModelContent(() => {
                const newValue = monacoEditor.current!.getValue();
                if (onChange) {
                    onChange(newValue, monacoEditor.current!);
                }
                handleShowPlaceholder(newValue);
            });

            monacoEditor.current.onDidBlurEditorText(() => {
                if (onBlur) {
                    const val = monacoEditor.current!.getValue();
                    onBlur(val);
                }
            });

            monacoEditor.current.onDidBlurEditorWidget(() => {
                handleShowPlaceholder(monacoEditor.current!.getValue());
            });

            monacoEditor.current.onDidFocusEditorText(() => {
                if (onFocus) {
                    const val = monacoEditor.current!.getValue();
                    onFocus(val);
                }
            });

            monacoEditor.current.onDidChangeCursorSelection(() => {
                const ranges = monacoEditor.current!.getSelections() || [];
                const model = monacoEditor.current!.getModel();
                let selectionContent = '';
                if (model) {
                    for (let i = 0; i < ranges.length; i += 1) {
                        selectionContent += model.getValueInRange(ranges[i]);
                    }
                }
                if (onCursorSelection) {
                    onCursorSelection(selectionContent);
                }
            });
            /**
             * 改变contextMenu的定位为fixed，避免容器内overflow:hidden属性截断contextMenu
             */
            monacoEditor.current.onContextMenu((e) => {
                const contextMenuElement = monacoEditor.current
                    ?.getDomNode()
                    ?.querySelector<HTMLElement>('.monaco-menu-container');

                if (contextMenuElement) {
                    const posY =
                        e.event.posy + contextMenuElement.clientHeight > window.innerHeight
                            ? e.event.posy - contextMenuElement.clientHeight
                            : e.event.posy;

                    const posX =
                        e.event.posx + contextMenuElement.clientWidth > window.innerWidth
                            ? e.event.posx - contextMenuElement.clientWidth
                            : e.event.posx;

                    contextMenuElement.style.position = 'fixed';
                    contextMenuElement.style.top = `${Math.max(0, Math.floor(posY))}px`;
                    contextMenuElement.style.left = `${Math.max(0, Math.floor(posX))}px`;
                }
            });
        }
    };

    const initEditor = () => {
        initEditorEvent();
    };

    const handleShowPlaceholder = (val?: string) => {
        if (!val) {
            placeholderDOM.current!.style.display = 'initial';
        } else {
            placeholderDOM.current!.style.display = 'none';
        }
    };

    const destroyMonaco = () => {
        if (monacoEditor.current) {
            monacoEditor.current.dispose();
        }
    };

    useEffect(() => {
        if (monacoEditor.current) {
            if (sync) {
                /**
                 * value更新， 并且含有sync同步标记，则更新编辑器值
                 */
                const editorText = value || '';
                monacoEditor.current.setValue(editorText);
                handleShowPlaceholder(editorText);

                const currentLanguage = monacoEditor.current.getModel()?.getLanguageId();
                const isScrollToBottom = currentLanguage?.endsWith('log');
                if (isScrollToBottom) {
                    monacoEditor.current.revealLineInCenterIfOutsideViewport(
                        monacoEditor.current.getModel()!.getLineCount()
                    );
                }
            }
        }
    }, [value]);

    useEffect(() => {
        if (monacoEditor.current) {
            monaco.editor.setModelLanguage(monacoEditor.current.getModel()!, language || 'ini');
        }
    }, [language]);

    useEffect(() => {
        if (monacoEditor.current && options) {
            monacoEditor.current.updateOptions(options);
        }
    }, [options]);

    useEffect(() => {
        initMonaco();

        return () => {
            destroyMonaco();
        };
    }, []);

    return (
        <div
            className={classNames('code-editor', className)}
            style={{
                position: 'relative',
                minHeight: '400px',
                height: '100%',
                width: '100%',
                ...style,
            }}
            ref={container}
        >
            <pre ref={placeholderDOM} className="dt-placeholder" style={{ fontSize: options?.fontSize || 12 }}>
                {placeholder}
            </pre>
        </div>
    );
}
