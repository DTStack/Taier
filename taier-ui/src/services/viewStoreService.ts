import { GlobalEvent } from '@dtinsight/molecule/esm/common/event';
import 'reflect-metadata';
import { singleton } from 'tsyringe';

enum ViewStorageEventKind {
    onStorageChange = 'onStorageChange',
}

// 负责整个应用的视图的持久化
@singleton()
class ViewStoreService extends GlobalEvent {
    private viewStorage = new Map<string, any>();

    public setViewStorage = <T>(tabId: string, value: ((preVal: T) => T) | T) => {
        if (typeof value === 'function') {
            this.viewStorage.set(tabId, (value as (preVal: T) => T)(this.getViewStorage(tabId)));
        } else {
            this.viewStorage.set(tabId, value);
        }
    };

    public getViewStorage = <T>(tabId: string) => {
        return <T>this.viewStorage.get(tabId);
    };

    /**
     * 删除 map 中存储的值, 若存在参数则删除某一个值，不存在则清空全部
     */
    public clearStorage = (tabId?: string) => {
        if (tabId === undefined) {
            this.viewStorage.clear();
            return true;
        }

        return this.viewStorage.delete(tabId);
    };

    public emiStorageChange(tabId: string) {
        this.emit(ViewStorageEventKind.onStorageChange, tabId);
    }

    public onStorageChange(callback: (tabId: string) => void) {
        this.subscribe(ViewStorageEventKind.onStorageChange, callback);
    }
}

export default new ViewStoreService();
