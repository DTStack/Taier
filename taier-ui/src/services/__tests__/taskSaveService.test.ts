import api from '@/api';
import { DATA_SOURCE_ENUM, FLINK_VERSIONS, KAFKA_DATA_TYPE, SOURCE_TIME_TYPE } from '@/constant';
import type { IOfflineTaskProps } from '@/interface';
import { isEditing } from '@/pages/editor/workflow';
import molecule from '@dtinsight/molecule';
import { waitFor } from '@testing-library/react';
import { message } from 'antd';
import { catalogueService, rightBarService } from '..';
import taskSaveService, { SaveEventKind } from '../taskSaveService';
import viewStoreService from '../viewStoreService';
import {
    guideAcquisition,
    guideFlinkSQL,
    guideSyncTabInWorkflow,
    scriptSyncTab,
    virtualTask,
} from './fixtures/taskSaveService';

jest.mock('@/api');
jest.mock('@/services/viewStoreService', () => ({
    getViewStorage: jest.fn(),
    emiStorageChange: jest.fn(),
}));
jest.mock('@/services', () => ({
    catalogueService: {
        onUpdate: jest.fn(),
        loadTreeNode: jest.fn(),
    },
    breadcrumbService: {
        getBreadcrumb: jest.fn(),
    },
    rightBarService: {
        getForm: jest.fn(() => ({
            componentForm: {
                validateFields: jest.fn(),
            },
        })),
    },
}));
jest.mock('antd', () => ({
    message: {
        success: jest.fn(),
        error: jest.fn(),
    },
}));

describe('Test TaskSaveService', () => {
    it('Should generate rules for source in flink', () => {
        expect(
            taskSaveService.generateValidDesSource(
                {
                    type: DATA_SOURCE_ENUM.KAFKA,
                    sourceDataType: KAFKA_DATA_TYPE.TYPE_AVRO,
                    timeTypeArr: [SOURCE_TIME_TYPE.EVENT_TIME],
                } as IOfflineTaskProps['source'][number],
                FLINK_VERSIONS.FLINK_1_12
            )
        ).toMatchSnapshot();
    });

    it('Should transform tab data', () => {
        expect(
            taskSaveService.transformTabDataToParams({
                value: 'this is value',
                componentVersion: FLINK_VERSIONS.FLINK_1_12,
                source: [
                    {
                        timeTypeArr: [1],
                    },
                ],
            } as unknown as IOfflineTaskProps)
        ).toEqual({
            value: 'this is value',
            sqlText: 'this is value',
            componentVersion: FLINK_VERSIONS.FLINK_1_12,
            source: [
                {
                    timeTypeArr: [1],
                    procTime: 'proc_time',
                },
            ],
        });
    });

    describe('Test save function', () => {
        beforeEach(() => {
            (message.error as jest.Mock).mockReset();
            (taskSaveService.emit as jest.Mock).mockReset();
            (api.saveTask as jest.Mock).mockReset().mockResolvedValue({
                code: 1,
                data: {},
            });
            (api.addOfflineTask as jest.Mock).mockReset().mockResolvedValue({
                code: 1,
                data: {},
            });
        });

        it('Should save sync tab with script mode', async () => {
            (api.saveOfflineJobData as jest.Mock).mockReset().mockResolvedValue({
                code: 1,
                data: {},
            });
            (molecule.editor.getState as jest.Mock).mockReset().mockImplementation(() => ({
                current: {
                    tab: scriptSyncTab,
                },
            }));

            // save script sync task at first time
            taskSaveService.save();

            await waitFor(() => {
                expect(api.saveOfflineJobData).toBeCalledTimes(1);
                expect(api.saveOfflineJobData).toBeCalledWith({
                    computeType: 1,
                    createModel: 1,
                    preSave: true,
                    settingMap: { channel: '1', speed: '-1' },
                    sourceMap: {
                        column: [
                            { index: 0, type: 'STRING' },
                            { index: 1, type: 'STRING' },
                        ],
                        encoding: 'utf-8',
                        fieldDelimiter: ',',
                        fileType: 'txt',
                        halfStructureDaType: 0,
                        isFirstLineHeader: true,
                        path: '/home/test.txt',
                        rdbmsDaType: 2,
                        sourceId: 23,
                        sourceList: [],
                        syncModel: 0,
                        type: 9,
                    },
                    sqlText: '',
                    targetMap: {
                        column: [
                            { comment: '', isPart: false, key: 'string', part: false, type: 'INT' },
                            { comment: '', isPart: false, key: 'age', part: false, type: 'INT' },
                        ],
                        sourceId: 17,
                        table: 'ftp_new',
                        type: 1,
                    },
                    taskType: 2,
                    test: 1,
                });
                // emit onSaveTask event handler after saving
                expect(taskSaveService.emit).toBeCalledTimes(1);
                expect(taskSaveService.emit).toBeCalledWith(SaveEventKind.onSaveTask, {});
            });
        });

        it('Should save sync tab with guide mode', async () => {
            (api.saveOfflineJobData as jest.Mock).mockReset().mockResolvedValue({
                code: 1,
                data: {},
            });

            (molecule.editor.getState as jest.Mock).mockReset().mockImplementation(() => ({
                current: {
                    tab: guideSyncTabInWorkflow,
                },
            }));

            (molecule.folderTree.get as jest.Mock).mockReset().mockImplementation(() => ({
                data: {
                    parentId: 1,
                },
            }));
            (taskSaveService.emit as jest.Mock)
                .mockReset()
                .mockImplementationOnce((_, ctx) => {
                    ctx.stop();
                })
                .mockImplementation((type, ctx) => {
                    if (type === SaveEventKind.onBeforeSave) {
                        ctx.continue();
                    }
                });

            // save guide sync task
            expect(taskSaveService.save()).rejects.toEqual(new Error('请检查数据同步任务是否填写正确'));

            await waitFor(() => {
                expect(taskSaveService.emit).toBeCalledTimes(1);
                expect((taskSaveService.emit as jest.Mock).mock.calls[0][0]).toBe(SaveEventKind.onBeforeSave);
            });

            expect(taskSaveService.save()).resolves.toEqual({
                code: 1,
                data: {},
            });

            await waitFor(() => {
                expect(taskSaveService.emit).toBeCalledTimes(3);
                expect(taskSaveService.emit).lastCalledWith(SaveEventKind.onSaveTask, {});
            });
        });

        it('Should check side failed in flinkSQL', async () => {
            const redisPrimaryKeyLost: typeof guideFlinkSQL = JSON.parse(JSON.stringify(guideFlinkSQL));
            redisPrimaryKeyLost.data.side[0].type = 12;
            (redisPrimaryKeyLost.data.side[0] as any).columnsText = '123';

            (molecule.editor.getState as jest.Mock).mockReset().mockImplementationOnce(() => ({
                current: {
                    tab: redisPrimaryKeyLost,
                },
            }));

            await expect(taskSaveService.save()).rejects.toEqual(new Error('维表1中的主键不能为空'));
        });

        it('Should save flink SQL in guide mode', async () => {
            const failedGuideFlinkSQL: typeof guideFlinkSQL = JSON.parse(JSON.stringify(guideFlinkSQL));
            failedGuideFlinkSQL.data.source[0].table = '';

            (molecule.editor.getState as jest.Mock)
                .mockReset()
                .mockImplementationOnce(() => ({
                    current: {
                        tab: failedGuideFlinkSQL,
                    },
                }))
                .mockImplementation(() => ({
                    current: {
                        tab: guideFlinkSQL,
                    },
                }));

            await expect(taskSaveService.save()).rejects.toEqual(undefined);

            expect((message.error as jest.Mock).mock.calls.map((call) => call.join('\n')).join('\n')).toBe(
                '源表 1 : 请输入映射表名'
            );

            await expect(taskSaveService.save()).resolves.toEqual({ code: 1, data: {} });

            expect(api.saveTask).toBeCalled();
            expect((api.saveTask as jest.Mock).mock.calls[0][0]).toMatchSnapshot();
            expect(taskSaveService.emit).toBeCalledWith(SaveEventKind.onSaveTask, {});
        });

        it('Should save flink SQL in script mode', async () => {
            const scriptFlinkSQL = JSON.parse(JSON.stringify(guideFlinkSQL));
            scriptFlinkSQL.data.createModel = 1;

            (molecule.editor.getState as jest.Mock).mockReset().mockImplementation(() => ({
                current: {
                    tab: scriptFlinkSQL,
                },
            }));

            await expect(taskSaveService.save()).resolves.toEqual({ code: 1, data: {} });
            expect(api.saveTask).toBeCalled();
            expect((api.saveTask as jest.Mock).mock.calls[0][0]).toMatchSnapshot();
        });

        it('Should save DATA_ACQUISITION failed', async () => {
            (rightBarService.getForm as jest.Mock)
                .mockReset()
                .mockImplementationOnce(() => ({
                    validateFields: jest.fn().mockRejectedValue(new Error('test')),
                }))
                .mockImplementation(() => ({
                    validateFields: jest.fn().mockResolvedValue({}),
                }));

            const failedAcquisition: typeof guideAcquisition = JSON.parse(JSON.stringify(guideAcquisition));
            failedAcquisition.data.sourceMap.pavingData = false;
            failedAcquisition.data.targetMap.type = 7;

            (molecule.editor.getState as jest.Mock).mockReset().mockImplementation(() => ({
                current: {
                    tab: failedAcquisition,
                },
            }));

            // failed at first time since validateFields failed;
            await expect(taskSaveService.save()).rejects.toEqual(new Error('test'));

            // failed at second time
            await expect(taskSaveService.save()).rejects.toEqual(undefined);
            expect(message.error).toBeCalledWith('请勾选嵌套Json平铺后重试');
        });

        it('Should save DATA_ACQUISITION successfully', async () => {
            (rightBarService.getForm as jest.Mock).mockReset().mockImplementation(() => ({
                validateFields: jest.fn().mockResolvedValue({}),
            }));
            (molecule.editor.getState as jest.Mock).mockReset().mockImplementation(() => ({
                current: {
                    tab: guideAcquisition,
                },
            }));

            await expect(taskSaveService.save()).resolves.toEqual({ code: 1, data: {} });

            expect(api.saveTask).toBeCalledTimes(1);
            expect((api.saveTask as jest.Mock).mock.calls[0][0]).toMatchSnapshot();
        });

        it('Should save virtual task', async () => {
            (molecule.folderTree.get as jest.Mock).mockReset().mockImplementation(() => ({
                data: {
                    parentId: 1,
                },
            }));

            (molecule.editor.getState as jest.Mock).mockReset().mockImplementation(() => ({
                current: {
                    tab: virtualTask,
                },
            }));

            await expect(taskSaveService.save()).resolves.toEqual({ code: 1, data: {} });

            expect(api.addOfflineTask).toBeCalledTimes(1);
            expect((api.addOfflineTask as jest.Mock).mock.calls[0][0]).toMatchSnapshot();

            expect((taskSaveService.emit as jest.Mock).mock.calls[0]).toEqual([SaveEventKind.onSaveTask, {}]);
        });

        it('Should save workflow task failed', async () => {
            (viewStoreService.getViewStorage as jest.Mock).mockReset().mockImplementation(() => ({
                cells: [
                    { edge: true },
                    {
                        vertex: true,
                        value: {
                            id: 'workflow__abc',
                            nodePid: 1,
                            [isEditing]: true,
                        },
                        setValue: jest.fn(),
                    },
                ],
            }));

            (api.addOfflineTask as jest.Mock).mockReset().mockResolvedValue({
                code: 0,
                data: {},
                message: 'test',
            });

            (molecule.editor.getState as jest.Mock).mockReset().mockImplementation(() => ({
                current: {
                    tab: {
                        data: {
                            id: 1,
                            taskType: 10,
                        },
                    },
                },
            }));

            await expect(taskSaveService.save()).rejects.toEqual(undefined);
            expect(message.error).lastCalledWith('test');
        });

        it('Should save workflow task successfully', async () => {
            (viewStoreService.getViewStorage as jest.Mock).mockReset().mockImplementation(() => ({
                cells: [
                    { edge: true, source: { value: { id: 1 } }, target: { value: { id: 2 } } },
                    {
                        vertex: true,
                        value: {
                            id: 'workflow__abc',
                            nodePid: 1,
                            [isEditing]: true,
                        },
                        setValue: jest.fn(),
                    },
                ],
            }));

            (api.getOfflineTaskByID as jest.Mock).mockReset().mockResolvedValue({
                code: 1,
                data: {},
            });
            (api.addOfflineTask as jest.Mock).mockReset().mockResolvedValue({
                code: 1,
                data: {
                    id: 1,
                },
            });

            (molecule.editor.getState as jest.Mock).mockReset().mockImplementation(() => ({
                current: {
                    tab: {
                        data: {
                            id: 1,
                            taskType: 10,
                        },
                    },
                },
            }));

            (molecule.editor.isOpened as jest.Mock).mockReset().mockImplementation(() => true);
            (molecule.editor.updateTab as jest.Mock).mockReset();

            await expect(taskSaveService.save()).resolves.toEqual({ code: 1, data: { id: 1 } });
            expect(api.addOfflineTask).toBeCalledTimes(2);
            expect(api.addOfflineTask).lastCalledWith({
                id: 1,
                nodeMap: { '2': [1], workflow__abc: [] },
                taskType: 10,
            });
            expect((taskSaveService.emit as jest.Mock).mock.calls[0]).toEqual([SaveEventKind.onSaveTask, { id: 1 }]);
        });

        it('Should save sparkMR successfully', async () => {
            (molecule.editor.getState as jest.Mock).mockReset().mockImplementation(() => ({
                current: {
                    tab: {
                        data: {
                            id: 1,
                            name: 'hadooooooopMR',
                            resourceIdList: [7],
                            mainClass: 'a.x',
                            exeArgs: null,
                            componentVersion: '2.1',
                            nodePid: 1,
                            taskType: 1,
                            taskDesc: '',
                            updateSource: false,
                            preSave: false,
                            test: 1,
                            scheduleConf:
                                '{"beginHour":"0","endHour":"23","beginMin":"0","gapHour":"5","periodType":"1","isFailRetry":true,"beginDate":"2001-01-01","endDate":"2121-01-01","selfReliance":0,"endMin":"59","maxRetryNum":3}',
                        },
                    },
                },
            }));

            const original = taskSaveService.updateFolderAndTabAfterSave;
            taskSaveService.updateFolderAndTabAfterSave = jest.fn();

            await expect(taskSaveService.save()).resolves.toEqual({ code: 1, data: {} });

            expect((api.addOfflineTask as jest.Mock).mock.calls[0][0]).toMatchSnapshot();
            expect((taskSaveService.emit as jest.Mock).mock.calls[0]).toEqual([SaveEventKind.onSaveTask, {}]);
            taskSaveService.updateFolderAndTabAfterSave = original;
        });

        it('Should save sparkSQL successful', async () => {
            (molecule.editor.getState as jest.Mock).mockReset().mockImplementation(() => ({
                current: {
                    tab: {
                        data: {
                            id: 'workflow__abc',
                            flowId: 1,
                            value: 'test',
                            nodePid: 1,
                            taskType: 0,
                        },
                    },
                },
            }));

            (api.saveOfflineJobData as jest.Mock).mockReset().mockResolvedValue({
                code: 1,
                data: {},
            });

            await expect(taskSaveService.save()).resolves.toEqual({ code: 1, data: {} });

            expect((api.saveOfflineJobData as jest.Mock).mock.calls[0][0]).toMatchSnapshot();
            expect((taskSaveService.emit as jest.Mock).mock.calls[0]).toEqual([SaveEventKind.onSaveTask, {}]);
        });
    });

    it('Should support save specific tab', async () => {
        (api.addOfflineTask as jest.Mock).mockReset().mockResolvedValue({
            code: 1,
            data: {
                id: 1,
                nodePid: 2,
            },
        });

        (api.getOfflineTaskByID as jest.Mock).mockReset().mockResolvedValue({
            code: 1,
            data: {
                flowId: 2,
            },
        });

        (molecule.editor.isOpened as jest.Mock).mockReset().mockImplementation(() => true);
        (molecule.editor.updateTab as jest.Mock).mockReset();
        (viewStoreService.getViewStorage as jest.Mock).mockReset().mockImplementation(() => ({
            cells: [
                {
                    vertex: true,
                    value: {
                        id: 1,
                    },
                    setValue: jest.fn(),
                },
            ],
        }));
        (viewStoreService.emiStorageChange as jest.Mock).mockReset();

        await expect(
            taskSaveService.saveTab({
                id: 1,
                nodePid: 2,
            } as IOfflineTaskProps)
        ).resolves.toEqual({ code: 1, data: { id: 1, nodePid: 2 } });

        expect((api.addOfflineTask as jest.Mock).mock.calls[0][0]).toMatchSnapshot();

        // set tab's status to saved
        expect(molecule.editor.updateTab).toBeCalledWith({
            id: '1',
            data: {
                flowId: 2,
            },
            status: undefined,
        });
    });

    it('Should update tab and folder after save', async () => {
        (molecule.folderTree.get as jest.Mock).mockReset().mockImplementation(() => ({}));
        (catalogueService.loadTreeNode as jest.Mock).mockReset();
        (molecule.editor.isOpened as jest.Mock).mockReset().mockImplementation(() => true);
        (molecule.editor.updateTab as jest.Mock).mockReset();
        (molecule.editor.getGroupIdByTab as jest.Mock).mockReset();
        (molecule.editor.closeTab as jest.Mock).mockReset();

        taskSaveService.updateFolderAndTabAfterSave(1, 2, 3, 'test', 'ttt');

        await waitFor(() => {
            expect(catalogueService.loadTreeNode).toBeCalledTimes(2);
            expect(molecule.editor.updateTab).toBeCalled();
            expect(molecule.editor.closeTab).toBeCalled();
        });
    });
});
