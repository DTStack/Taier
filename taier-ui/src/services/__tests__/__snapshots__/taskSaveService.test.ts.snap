// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test TaskSaveService Should generate rules for source in flink 1`] = `
Object {
  "columnsText": Array [
    Object {
      "message": "字段信息不能为空！",
      "required": true,
    },
  ],
  "offset": Array [
    Object {
      "message": "请输入最大延迟时间",
      "required": true,
    },
  ],
  "schemaInfo": Array [
    Object {
      "message": "请输入Schema",
      "required": false,
    },
  ],
  "sourceDataType": Array [
    Object {
      "message": "请选择读取类型",
      "required": true,
    },
  ],
  "sourceId": Array [
    Object {
      "message": "请选择数据源",
      "required": true,
    },
  ],
  "table": Array [
    Object {
      "message": "请输入映射表名",
      "required": true,
    },
  ],
  "timeColumn": Array [
    Object {
      "message": "请选择时间列",
      "required": true,
    },
  ],
  "topic": Array [
    Object {
      "message": "请选择Topic",
      "required": true,
    },
  ],
  "type": Array [
    Object {
      "message": "请选择类型",
      "required": true,
    },
  ],
}
`;

exports[`Test TaskSaveService Should support save specific tab 1`] = `
Object {
  "computeType": 1,
  "id": 1,
  "nodePid": 2,
}
`;

exports[`Test TaskSaveService Test save function Should save DATA_ACQUISITION successfully 1`] = `
Object {
  "createModel": 0,
  "preSave": true,
  "settingMap": Object {
    "readerChannel": "1",
    "speed": -1,
    "writerChannel": 1,
  },
  "sourceMap": Object {
    "allTable": false,
    "cat": Array [
      1,
      2,
      3,
    ],
    "codec": "plain",
    "codecType": "text",
    "collectPoint": "taskRun",
    "collectType": 0,
    "decode": "text",
    "decoder": "FixedLength",
    "distributeTable": Object {
      "test": Array [
        "a",
      ],
    },
    "isCleanSession": true,
    "mode": "group-offsets",
    "multipleTable": false,
    "parse": "text",
    "pavingData": false,
    "protocol": "http",
    "qos": 2,
    "rdbmsDaType": 1,
    "requestMode": "post",
    "slotConfig": 1,
    "sourceId": 17,
    "table": Array [
      "binary_test",
    ],
    "temporary": false,
    "type": 1,
  },
  "sqlText": "",
  "targetMap": Object {
    "dataSequence": false,
    "isCleanSession": true,
    "qos": 2,
    "sourceId": 19,
    "topic": "test",
    "type": 37,
  },
  "taskType": 6,
  "updateSource": true,
}
`;

exports[`Test TaskSaveService Test save function Should save flink SQL in guide mode 1`] = `
Object {
  "componentVersion": "1.12",
  "createModel": 0,
  "preSave": true,
  "side": Array [
    Object {
      "asyncPoolSize": 5,
      "cache": "LRU",
      "cacheSize": 10000,
      "cacheTTLMs": 60000,
      "columns": Array [
        Object {
          "column": "i1",
          "type": "VARBINARY",
        },
      ],
      "parallelism": 1,
      "sourceId": 9,
      "table": "binary_test",
      "tableName": "test",
      "type": 1,
    },
  ],
  "sink": Array [
    Object {
      "allReplace": "false",
      "batchSize": 100,
      "batchWaitInterval": 1000,
      "bulkFlushMaxActions": 100,
      "columns": Array [
        Object {
          "column": "i1",
          "type": "VARBINARY",
        },
      ],
      "enableKeyPartitions": false,
      "parallelism": 1,
      "sourceId": 9,
      "table": "binary_test",
      "tableName": "test",
      "type": 1,
      "updateMode": "append",
    },
  ],
  "source": Array [
    Object {
      "charset": "utf-8",
      "columnsText": "id int",
      "offset": 0,
      "offsetReset": "latest",
      "offsetUnit": "SECOND",
      "parallelism": 1,
      "procTime": "proc_time",
      "sourceDataType": "dt_nest",
      "sourceId": "19",
      "table": "test",
      "timeType": 1,
      "timeTypeArr": Array [
        1,
      ],
      "timeZone": "Asia/Shanghai",
      "topic": "test_topic_name_gcwfybte",
      "type": 37,
    },
  ],
  "sqlText": "",
  "taskType": 5,
  "updateSource": true,
}
`;

exports[`Test TaskSaveService Test save function Should save flink SQL in script mode 1`] = `
Object {
  "componentVersion": "1.12",
  "createModel": 1,
  "preSave": true,
  "side": Array [
    Object {
      "asyncPoolSize": 5,
      "cache": "LRU",
      "cacheSize": 10000,
      "cacheTTLMs": 60000,
      "columns": Array [
        Object {
          "column": "i1",
          "type": "VARBINARY",
        },
      ],
      "parallelism": 1,
      "sourceId": 9,
      "table": "binary_test",
      "tableName": "test",
      "type": 1,
    },
  ],
  "sink": Array [
    Object {
      "allReplace": "false",
      "batchSize": 100,
      "batchWaitInterval": 1000,
      "bulkFlushMaxActions": 100,
      "columns": Array [
        Object {
          "column": "i1",
          "type": "VARBINARY",
        },
      ],
      "enableKeyPartitions": false,
      "parallelism": 1,
      "sourceId": 9,
      "table": "binary_test",
      "tableName": "test",
      "type": 1,
      "updateMode": "append",
    },
  ],
  "source": Array [
    Object {
      "charset": "utf-8",
      "columnsText": "id int",
      "offset": 0,
      "offsetReset": "latest",
      "offsetUnit": "SECOND",
      "parallelism": 1,
      "sourceDataType": "dt_nest",
      "sourceId": "19",
      "table": "test",
      "timeType": 1,
      "timeTypeArr": Array [
        1,
      ],
      "timeZone": "Asia/Shanghai",
      "topic": "test_topic_name_gcwfybte",
      "type": 37,
    },
  ],
  "sqlText": undefined,
  "taskType": 5,
  "updateSource": true,
}
`;

exports[`Test TaskSaveService Test save function Should save sparkMR successfully 1`] = `
Object {
  "componentVersion": "2.1",
  "exeArgs": null,
  "id": 1,
  "mainClass": "a.x",
  "name": "hadooooooopMR",
  "nodePid": 1,
  "preSave": false,
  "resourceIdList": Array [
    7,
  ],
  "scheduleConf": "{\\"beginHour\\":\\"0\\",\\"endHour\\":\\"23\\",\\"beginMin\\":\\"0\\",\\"gapHour\\":\\"5\\",\\"periodType\\":\\"1\\",\\"isFailRetry\\":true,\\"beginDate\\":\\"2001-01-01\\",\\"endDate\\":\\"2121-01-01\\",\\"selfReliance\\":0,\\"endMin\\":\\"59\\",\\"maxRetryNum\\":3}",
  "taskDesc": "",
  "taskType": 1,
  "test": 1,
  "updateSource": false,
}
`;

exports[`Test TaskSaveService Test save function Should save sparkSQL successful 1`] = `
Object {
  "computeType": 1,
  "flowId": 1,
  "nodePid": 1,
  "preSave": true,
  "sqlText": "test",
  "taskType": 0,
}
`;

exports[`Test TaskSaveService Test save function Should save virtual task 1`] = `
Object {
  "computeType": 1,
  "flowId": 1,
  "name": "virtual",
  "nodePid": 1,
  "taskDesc": "",
  "taskType": -1,
}
`;
